libopenage/game_control.cpp

```diff
@@ -555,83 +382,209 @@ bool EditorMode::on_single_click(int del, coord::window point) {
-	this->modes.push_back(std::make_unique<EditorMode>());
-	this->modes.push_back(std::make_unique<ActionMode>());
+void EditorMode::announce_category_content(const std::string &category_name) {
+	if (this->game_control)
+		if (auto player = this->game_control->get_current_player()) {
+			auto inds = player->civ->get_category(category_name);
+			std::vector<std::tuple<index_t, uint16_t>> type_and_texture(inds.size());
 
-	// initial active mode
-	this->active_mode = modes.front().get();
-	engine->get_input_manager().register_context(this->active_mode);
+			auto it = std::begin(type_and_texture);
 
-	auto &global_input_context = engine->get_input_manager().get_global_context();
-	global_input_context.bind(input::actions::TOGGLE_CONSTRUCT_MODE, [this](const input::action_arg_t &) {
-		this->toggle_mode();
-	});
+			std::for_each(std::begin(inds), std::end(inds), [player, &it] (auto index) {
+				*it++ = std::make_tuple(index, player->get_type(index)->default_texture()->id);
+			});
 
+			emit this->gui_signals.category_content_changed(category_name, type_and_texture);
+		}
+}
+
+void EditorMode::announce() {
+	OutputMode::announce();
+	announce_categories();
 }
 
-void GameControl::toggle_mode() {
-	int next_mode = (this->active_mode_index + 1) % this->modes.size();
-	if (this->modes[next_mode]->available()) {
-		engine->get_input_manager().remove_context(this->active_mode);
+void EditorMode::set_game_control(GameControl *game_control) {
+	if (this->game_control != game_control) {
+		if (this->game_control)
+			QObject::disconnect(&this->game_control->gui_signals, &GameControlSignals::current_player_name_changed, &this->gui_signals, &EditorModeSignals::on_current_player_name_changed);
 
-		// set the new active mode
-		this->active_mode_index = next_mode;
-		this->active_mode = this->modes[next_mode].get();
-		this->active_mode->on_enter();
+		OutputMode::set_game_control(game_control);
 
-		// update the context
-		engine->get_input_manager().register_context(this->active_mode);
+		if (this->game_control)
+			QObject::connect(&this->game_control->gui_signals, &GameControlSignals::current_player_name_changed, &this->gui_signals, &EditorModeSignals::on_current_player_name_changed);
+
+		announce_categories();
+	} else {
+		OutputMode::set_game_control(game_control);
 	}
 }
 
-bool GameControl::on_drawhud() {
-	Engine &engine = Engine::get();
-	int x = engine.engine_coord_data->window_size.x - 300;
-	int y = engine.engine_coord_data->window_size.y - 24;
-	engine.render_text({x, y}, 20, "%s", active_mode->name().c_str());
+GameControlSignals::GameControlSignals(GameControl *game_control)
+	:
+	game_control{game_control} {
+}
+
+void GameControlSignals::on_game_running(bool running) {
+	if (running)
+		this->game_control->announce_current_player_name();
+}
+
+GameControl::GameControl(gui::GuiItemLink *gui_link)
+	:
+	engine{},
+	game{},
+	active_mode{},
+	active_mode_index{-1},
+	current_player{1},
+	gui_signals{this},
+	gui_link{gui_link} {
+}
```

Comment:
That's related to the QML fundamental lack of constructors (because the entire declaration of the object in the script is something like a constructor).

The script engine creates C++ objects in no particular order and with default values. Then it calls setters on the objects according to the values that are defined in the script for their members.